lib/data-types.js

```diff
@@ -1,281 +1,484 @@
-  } else {
-    return new CHAR(length, binary);
-  }
+ABSTRACT.prototype.toString = function() {
+  return this.toSql();
+};
+ABSTRACT.prototype.toSql = function() {
+  return this.key;
+};
+
+/**
+ * A variable length string. Default length 255
+ *
+ * Available properties: `BINARY`
+ *
+ * @property STRING
+ */
+var STRING = function(length, binary) {
+  var options = typeof length === "object" && length || {
+    length: length,
+    binary: binary
+  };
+
+  if (!(this instanceof STRING)) return new STRING(options);
+
+  this._binary = options.binary;
+  this._length = options.length || 255;
 };
+util.inherits(STRING, ABSTRACT);
 
-STRING.prototype = {
-  get BINARY() {
+STRING.prototype.key = 'STRING';
+STRING.prototype.toSql = function() {
+  return 'VARCHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
+};
+Object.defineProperty(STRING.prototype, 'BINARY', {
+  get: function() {
     this._binary = true;
     return this;
-  },
-  get type() {
-    return this.toString();
-  },
-  toString: function() {
-    return 'VARCHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
   }
+});
+
+/**
+ * A fixed length string. Default length 255
+ *
+ * Available properties: `BINARY`
+ *
+ * @property CHAR
+ */
+var CHAR = function(length, binary) {
+  var options = typeof length === "object" && length || {
+    length: length,
+    binary: binary
+  };
+
+  if (!(this instanceof CHAR)) return new CHAR(options);
+  STRING.apply(this, arguments);
 };
+util.inherits(CHAR, STRING);
 
-CHAR.prototype = {
-  get BINARY() {
-    this._binary = true;
-    return this;
-  },
-  get type() {
-    return this.toString();
-  },
-  toString: function() {
-    return 'CHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
+CHAR.prototype.key = 'CHAR';
+CHAR.prototype.toSql = function() {
+  return 'CHAR(' + this._length + ')' + ((this._binary) ? ' BINARY' : '');
+};
+
+
+/**
+ * An unlimited length text column
+ * @property TEXT
+ */
+var TEXT = function(options) {
+  if (!(this instanceof TEXT)) return new TEXT(options);
+};
+util.inherits(TEXT, ABSTRACT);
+
+TEXT.prototype.key = 'TEXT';
+
+
+var NUMBER = function(options) {
+  if (!(this instanceof NUMBER)) return new NUMBER(options);
+  this._length = options.length;
+  this._decimals = options.decimals;
+  this._precision = options.precision;
+  this._scale = options.scale;
+  this._unsigned = options.unsigned;
+};
+util.inherits(NUMBER, ABSTRACT);
+
+NUMBER.prototype.key = 'NUMBER';
+NUMBER.prototype.toSql = function() {
+  var result = this.key;
+  if (this._length) {
+    result += '(' + this._length;
+    if (typeof this._decimals === 'number') {
+      result += ',' + this._decimals;
+    }
+    result += ')';
+  }
+  if (this._unsigned) {
+    result += ' UNSIGNED';
+  }
+  if (this._zerofill) {
+    result += ' ZEROFILL';
   }
+  return result;
 };
 
-Object.defineProperty(STRING, 'BINARY', {
+Object.defineProperty(NUMBER.prototype, 'UNSIGNED', {
   get: function() {
```

Comment:
God i hate this, but i don't see how we provide BC otherwise. We can't even create a property that maps to a method since the prop names will clash.